### PR TITLE
Add vertical and horizontal vane position controls

### PR DESCRIPTION
**Pull request by nathanjw**
- Add vertical and horizontal vane selects as done for Panasonic pumps; https://github.com/DomiStyle/esphome-panasonic-ac
- Add HORIZONTAL and BOTH swing mode options in additon to OFF and VERTICAL
- Fix logging bug causing ESP to crash

Starting from https://github.com/geoffdavis/esphome-mitsubishiheatpump/pull/103 by @seime, extract the vane implementation code and format according to existing style. Leave behind all the reformatting.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -225,6 +225,10 @@ climate:
     # logging:baud_rate above to allow the built-in UART0 to function for
     # logging.
     hardware_uart: UART0
+    horizontal_vane_select:
+      name: Horizontal Vane Position
```
Comment (SSMCD) [inline README.md:229]: name: ${name} Horizontal Vane

File: README.md [hunk 2]
```diff
@@ -225,6 +225,10 @@ climate:
...
+    vertical_vane_select:
+      name: Horizontal Vane Position
```
Comment (SSMCD) [inline README.md:231]: name: ${name} Vertical Vane

**Review by SSMCD — COMMENTED** [review]
Confirmed working with my MITMSZFH09NA unit.   Thank you!

Recommend disabling the swing mode trait of the climate component if these Horizontal and Vertical selects are used.
See     [Advanced Configuration](https://github.com/geoffdavis/esphome-mitsubishiheatpump/blob/master/README.md#advanced-configuration):
```
supports:
  swing_mode: []
```

**Comment by swannet** [comment]
I've tested this PR on my units (MSZ-AP**VGK) and they are working as expected. I've used the by SSMCD on 4th august (line 231 > Horizontal Vane Position > ${name} Vertical Vane)

@geoffdavis sorry to ping you, would you please have a look af this PR to add the functionality to control the vanes?

**Comment by geoffdavis** [comment]
Can you rebase this onto devel instead of master? I've got some pending commits on the devel branch that I need to package as a release. They're mostly documentation changes.

File: README.md
[the remark this answers is not in the document]
Comment (geoffdavis) [reply README.md:1]: This likely went away when I set the base branch to develop. Can you work with @nathanjw on that rebase?

**Comment by nathanjw** [comment]
Rebased against devel, and caught up to the changes there. I also re-added the models that @seime mentioned.

**Review by geoffdavis — CHANGES_REQUESTED** [review]
Most units don't support Horizontal or Both swing modes _by default_. Please leave the default set to ['OFF', 'VERTICAL']

File: README.md [hunk 3]
```diff
@@ -380,7 +391,7 @@ climate:
   ** `fan_mode` (_Optional_, list): Supported fan speeds for the HeatPump.
     Default: `['AUTO', 'DIFFUSE', 'LOW', 'MEDIUM', 'MIDDLE', 'HIGH']`
   ** `swing_mode` (_Optional_, list): Supported fan swing modes. Most Mitsubishi
-    units only support the default. Default: `['OFF', 'VERTICAL']`
+    units only support the default. Default: `['OFF', 'VERTICAL', 'HORIZONTAL', 'BOTH']`
```
Comment (geoffdavis) [inline README.md:394]: The default should be ['OFF', 'VERTICAL']

File: components/mitsubishi_heatpump/climate.py [hunk 4]
```diff
@@ -15,12 +15,24 @@
...
-AUTO_LOAD = ["climate"]
+AUTO_LOAD = ["climate", "select"]
 
 CONF_SUPPORTS = "supports"
+CONF_HORIZONTAL_SWING_SELECT = "horizontal_vane_select"
+CONF_VERTICAL_SWING_SELECT = "vertical_vane_select"
 DEFAULT_CLIMATE_MODES = ["HEAT_COOL", "COOL", "HEAT", "DRY", "FAN_ONLY"]
 DEFAULT_FAN_MODES = ["AUTO", "DIFFUSE", "LOW", "MEDIUM", "MIDDLE", "HIGH"]
-DEFAULT_SWING_MODES = ["OFF", "VERTICAL"]
+DEFAULT_SWING_MODES = ["OFF", "VERTICAL", "HORIZONTAL", "BOTH"]
```
Comment (geoffdavis) [inline components/mitsubishi_heatpump/climate.py:25]: Keep this as ["OFF", "VERTICAL"]

Comment (nathanjw) [reply components/mitsubishi_heatpump/climate.py]: Done.

**Comment by SGXander** [comment]
Tested nathanjw:unformat-select and it's working well on MSZ-LN25VG2

**Comment by geoffdavis** [comment]
> Can this be merged? I Want to get rid of the MELCLOUD integration, but I need this kind of control to be able to do that..

Not until the [change I requested is made](https://github.com/geoffdavis/esphome-mitsubishiheatpump/pull/109#pullrequestreview-1677356783), and the conflicts are resolved.

**Comment by SGXander** [comment]
I've made a slight alteration at https://github.com/SGXander/esphome-mitsubishiheatpump/tree/split_fix which I'd be happy to PR either here or into nathanjws fork, let me know which. The underlying swicago library specifies "<>" which in nathanjws version is marked as "auto" but during testing I have discovered it is actually "split" (leaving the middle clear and sending air left and right). I've updated this in my fork under the split_fix branch but so far I haven't found the command code to activate auto on the horizontal_vane so have left it out of the options variable.

**Comment by nathanjw** [comment]
Please take another look. I had adjusted the defaults as requested, but missed that merge conflict. I think that is resolved now.

**Comment by eclair4151** [comment]
Whats the best way to test this PR on my Home Assistant install? I tried replacing the external component URL with your branch URL, but it didnt seem to have any effect.

**Comment by SGXander** [comment]
> Whats the best way to test this PR on my Home Assistant install? I tried replacing the external component URL with your branch URL, but it didnt seem to have any effect.

`external_components:
  - source: github://nathanjw/esphome-mitsubishiheatpump@unformat-select
`

**Comment by eclair4151** [comment]
Awesome, thanks! Just tested it out, working for me as well. I wish Home assistant let you customize the HVAC card to allow you to put the swing in the card, oh well :). @SGXander I think you still need to mark the PR ready for review.

**Comment by SGXander** [comment]
> Awesome, thanks! Just tested it out, working for me as well. I wish Home assistant let you customize the HVAC card to allow you to put the swing in the card, oh well :). @SGXander I think you still need to mark the PR ready for review. 

Not my repo, it's ready to merge already though if you want it to report "split" instead of "auto" then use my fork for now as I don't think nathanjw has updated this one.

**Comment by eclair4151** [comment]
Ah whoops, @nathanjw thoughts on adding in support for horizontal vane split position in this PR and marking it ready for review?

**Comment by SGXander** [comment]
put in https://github.com/nathanjw/esphome-mitsubishiheatpump/pull/1 for the split thing that way it should update this PR if approved

**Comment by eclair4151** [comment]
@nathanjw Whats the status on this? If we dont hear back should we just raise a separate PR based on yours SGXander which includes the fix?

**Comment by nathanjw** [comment]
I think I'd like to get this in as is and then a cosmetic renaming of that horizontal mode could be a much easier followon.

**Comment by jamiebegin** [comment]
> > Whats the best way to test this PR on my Home Assistant install? I tried replacing the external component URL with your branch URL, but it didnt seem to have any effect.
> 
> ```
> external_components:
>   - source: github://nathanjw/esphome-mitsubishiheatpump@unformat-select
> ```

I'm getting a whole slew of errors when I attempt to compile using this branch. It starts with:
```
In file included from src/esphome.h:33,
                 from src/esphome/components/mitsubishi_heatpump/espmhp.h:20,
                 from src/esphome/components/mitsubishi_heatpump/espmhp.cpp:21:
src/esphome/components/mitsubishi_heatpump/espmhp.h:182:14: error: 'optional' in namespace 'std' does not name a template type
         std::optional<std::chrono::duration<long long, std::ratio<60>>> remote_operating_timeout_;
```

These errors don't happen when using the base branch. I'm not much of a C++ guy so I'm not sure where to start looking for the problem?

**Comment by swannet** [comment]
@jamiebegin 

You have to add some options to help the compiler use C++17

See the sample below. You've to add the platformio_options with the build_unflag and build_flags

```
esphome:
  name: ${name}
  friendly_name: ${friendly_name}
  # Set platform options to use c++ 17 instead of 11
  platformio_options:
    build_unflags:
      - "-std=gnu++11"
    build_flags:
      - "-std=gnu++17"

```

**Comment by davidcd13** [comment]
Vane selection cannot be integrated into the default home assistant climate card? Is it necessarily a separate card? With Melcloud,  vane position is integrated into the default climate card.

**Comment by swannet** [comment]
> Vane selection cannot be integrated into the default home assistant climate card? Is it necessarily a separate card? With Melcloud, vane position is integrated into the default climate card.

It is possible, with the "features" you can add to the climate card. And with the service-call-tile-feature add-on from Nerwyn (https://github.com/Nerwyn/service-call-tile-feature):

Screenshot 2024-03-16 at 17 30 11

```
type: thermostat
entity: climate.airco_
features:
  - style: icons
    hvac_modes:
      - 'off'
      - cool
      - heat
      - fan_only
      - dry
    type: climate-hvac-modes
  - style: dropdown
    fan_modes:
      - auto
      - low
      - medium
      - high
    type: climate-fan-modes
  - type: custom:service-call
    entries:
      - type: selector
        entity_id: select.airco_woonkamer_horizontal_vane_position
        options:
          - option: left
            icon: mdi:chevron-double-left
          - option: left_center
            icon: mdi:chevron-left
          - option: center
            icon: mdi:format-horizontal-align-center
          - option: right_center
            icon: mdi:chevron-right
          - option: right
            icon: mdi:chevron-double-left
  - type: custom:service-call
    entries:
      - type: selector
        entity_id: select.airco_woonkamer_vertical_vane_position
        options:
          - option: up
            icon: mdi:chevron-double-up
          - option: up_center
            icon: mdi:chevron-up
          - option: center
            icon: mdi:format-vertical-align-center
          - option: down_center
            icon: mdi:chevron-down
          - option: down
            icon: mdi:chevron-double-down
show_current_as_primary: false
```

**Review by geoffdavis — APPROVED** [review]

**Pull request merged**